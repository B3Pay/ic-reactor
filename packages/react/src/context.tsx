import React, {
  createContext,
  useState,
  useEffect,
  PropsWithChildren,
  useMemo,
  useCallback,
} from "react"
import { ActorSubclass, HttpAgent } from "@dfinity/agent"
import { createReActor } from "./index"
import { IDL } from "@dfinity/candid"
import { getDidJsFromMetadata, getDidJsFromTmpHack } from "@ic-reactor/store"
import { ReActorOptions } from "@ic-reactor/store"

export type ReActor<A> = ReturnType<typeof createReActor<A>>

export interface ReActorContextType<A = ActorSubclass<any>>
  extends ReActor<A> {}

const ReActorContext = createContext<ReActorContextType | null>(null)

export const useReActor = <
  A = ActorSubclass<any>,
>(): ReActorContextType<A> => {
  const context = React.useContext(ReActorContext)

  if (!context) {
    throw new Error("useReActor must be used within a ReActorProvider")
  }

  return context
}

interface ReActorProviderProps
  extends PropsWithChildren,
    Omit<ReActorOptions, "idlFactory"> {
  idlFactory?: IDL.InterfaceFactory
  loadingComponent?: React.ReactNode
}

export const ReActorProvider: React.FC<ReActorProviderProps> = ({
  children,
  loadingComponent = <div>Loading...</div>,
  ...config
}) => {
  const [didJs, setDidJS] = useState<{ idlFactory: IDL.InterfaceFactory }>()

  const fetchDidJs = useCallback(async () => {
    const { canisterId } = config
    if (!canisterId) {
      throw new Error("canisterId is required")
    }

    const agent = new HttpAgent({ ...config })

    getDidJsFromMetadata(agent, canisterId).then(async (idlFactory) => {
      if (!idlFactory) {
        try {
          idlFactory = await getDidJsFromTmpHack(agent, canisterId)
        } catch (err) {
          if (/no query method/.test(err as any)) {
            console.warn(err)
            idlFactory = undefined
          } else {
            throw err
          }
        }

        if (!idlFactory) {
          console.warn("No query method found for canister", canisterId)
        }
      }
      setDidJS(idlFactory)
    })

    return didJs
  }, [config.canisterId])

  useEffect(() => {
    const { idlFactory } = config

    if (idlFactory) {
      setDidJS({ idlFactory })
      return
    }

    fetchDidJs()
  }, [fetchDidJs, config.idlFactory])

  const reActorHooks = useMemo(() => {
    if (!didJs) {
      return null
    }
    try {
      return createReActor<any>({
        ...config,
        idlFactory: didJs.idlFactory,
      })
    } catch (err) {
      console.error(err)
      return null
    }
  }, [config, didJs])

  return (
    <ReActorContext.Provider value={reActorHooks}>
      {reActorHooks ? children : loadingComponent}
    </ReActorContext.Provider>
  )
}
