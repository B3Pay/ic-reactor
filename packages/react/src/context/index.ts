export * from "./adapter"
export * from "./actor"
export * from "./agent"
