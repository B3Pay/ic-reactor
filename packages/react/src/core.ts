export {
  createReactorStore,
  createActorManager,
  createAgentManager,
  createCandidAdapter,
} from "@ic-reactor/core"
