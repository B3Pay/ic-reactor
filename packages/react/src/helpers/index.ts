export * from "./getAgent"
export * from "./getActor"
export * from "./getAuth"
export * from "./extractActor"
export * from "./extractAgent"
