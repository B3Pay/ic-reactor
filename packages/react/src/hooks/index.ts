export * from "./useReactor"
export * from "./useCandid"
