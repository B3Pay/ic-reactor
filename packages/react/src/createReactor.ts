import { createReactorStore } from "@ic-reactor/core"
import { agentHooks, authHooks, actorHooks } from "./helpers"

import type {
  BaseActor,
  CreateReactorParameters,
  CreateReactorReturnType,
} from "./types"

/**
 * Initializes and configures the reactor environment for interacting with the Internet Computer (IC) blockchain within a React application.
 * It encapsulates the creation of actors, authentication, and agent management, offering a streamlined interface for blockchain interactions.
 *
 * @param config Configuration config for the reactor, including:
 *  - `withProcessEnv` (optional): Specifies whether to use process environment variables to determine if the environment is local or development. Defaults to false.
 *  - `isLocalEnv` (optional): Indicates if the current environment is local or development, influencing the agent and actor behavior. Useful for testing or development.
 *  - `port` (optional): Port number for the local or development environment.
 *  Extends `CreateReactorStoreParameters` which includes HTTP agent config, actor manager config (excluding `agentManager`), and an optional custom agent manager.
 *
 * @returns An object containing various hooks and utilities:
 *  - {@link getAgent} - Returns the current agent instance.
 *  - {@link getVisitFunction} - Returns the visit function for the actor.
 *  - {@link useQueryCall} - A hook for querying actor methods.
 *  - {@link useUpdateCall} - A hook for updating actor methods.
 *  - {@link useAuth} - A hook for managing authentication and user principal.
 *  - {@link useActor} - A hook for managing actors and their methods.
 *  - {@link useActorManager} - A hook for managing actor manager and its methods.
 *  - {@link useAgentManager} - A hook for managing agent manager and its methods.
 *  - {@link initialize} - A function to initialize the actor manager if not initialized.
 *  - {@link useActorState} - A hook for managing actor state.
 *  - {@link useAgent} - A hook for managing agent and its methods.
 *  - {@link useAuthState} - A hook for managing authentication state.
 *  - {@link useAgentState} - A hook for managing agent state.
 *  - {@link useUserPrincipal} - A hook for managing user principal.
 *  - {@link useVisitMethod} - A hook for visiting actor methods.
 *
 * @example
 * ```typescript
 * import { createReactor } from "@ic-reactor/react";
 * import type { CreateReactorParameters } from "@ic-reactor/react/dist/types";
 * import { canisterId, idlFactory, yourActor } from "./declaration/yourActor"
 *
 * const config: CreateReactorParameters = {
 *   canisterId,
 *   idlFactory,
 *   host: "https://localhost:8000", // IC network host         |
 *   isLocalEnv: true, // Set true for local network            | one of these
 *   withProcessEnv: true, // Use process.env to determine host |
 *   port: 8000, // Port number for local network               |
 * };
 *
 * export type YourActor = typeof yourActor;
 *
 * export const { useAuth, useQueryCall, useUpdateCall } = createReactor<YourActor>(config);
 * // Now you can use the returned hooks in your React components
 * ```
 */
export const createReactor = <A = BaseActor>(
  config: CreateReactorParameters
): CreateReactorReturnType<A> => {
  const actorManager = createReactorStore<A>(config)

  const getVisitFunction = () => {
    return actorManager.visitFunction
  }

  const getAgent = () => {
    return actorManager.agentManager.getAgent()
  }

  return {
    getAgent,
    getVisitFunction,
    ...actorHooks(actorManager),
    ...authHooks(actorManager.agentManager),
    ...agentHooks(actorManager.agentManager),
  }
}
