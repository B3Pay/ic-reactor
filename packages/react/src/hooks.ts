import {
  ActorSubclass,
  CallMethod,
  ExtractReActorMethodArgs,
  ReActorActorStore,
} from "@ic-reactor/store"
import { useCallback, useEffect, useMemo, useRef, useState } from "react"
import {
  ReActorCallArgs,
  ReActorHookState,
  ReActorUseQueryArgs,
  ReActorUseUpdateArgs,
} from "./types"
import { useStore } from "zustand"

export const getCallHooks = <A extends ActorSubclass<any>>(
  callMethod: CallMethod<A>,
  actorStore: ReActorActorStore<A>
) => {
  const useMethodFields = () => {
    return useStore(actorStore, (state) => state.methodFields)
  }

  const useMethodField = (functionName: keyof A & string) => {
    const methodFields = useMethodFields()

    const field = useMemo(() => {
      return methodFields.find((f) => f.functionName === functionName)
    }, [methodFields, functionName])

    return field
  }

  const useReActorCall = <M extends keyof A>({
    onError,
    onSuccess,
    onLoading,
    args = [] as unknown as ExtractReActorMethodArgs<A[M]>,
    functionName,
  }: ReActorCallArgs<A, M>) => {
    const [state, setState] = useState<ReActorHookState<A, M>>({
      data: undefined,
      error: undefined,
      loading: false,
    })

    const call = useCallback(
      async (replaceArgs?: ExtractReActorMethodArgs<A[M]>) => {
        onLoading?.(true)
        onError?.(undefined)
        setState((prevState) => ({
          ...prevState,
          loading: true,
          error: undefined,
        }))

        try {
          const data = await callMethod(functionName, ...(replaceArgs ?? args))

          onLoading?.(false)
          onSuccess?.(data)
          setState((prevState) => ({ ...prevState, data, loading: false }))

          return data
        } catch (error) {
          console.error("Error in call:", error)
          onError?.(error as Error)
          onLoading?.(false)
          setState((prevState) => ({
            ...prevState,
            error: error as Error,
            loading: false,
          }))
        }
      },
      [args, functionName, onError, onSuccess, onLoading]
    )

    const field = useMethodField(functionName)

    return { call, field, ...state }
  }

  const useQueryCall = <M extends keyof A>({
    autoRefresh,
    refreshInterval = 5000,
    disableInitialCall,
    ...rest
  }: ReActorUseQueryArgs<A, M>) => {
    const { call, ...state } = useReActorCall(rest)

    let intervalId = useRef<NodeJS.Timeout | undefined>(undefined)

    useEffect(() => {
      // Auto-refresh logic
      if (autoRefresh) {
        intervalId.current = setInterval(() => {
          call()
        }, refreshInterval)
      }

      // Initial call logic
      if (!disableInitialCall) {
        call()
      }

      return () => {
        clearInterval(intervalId.current)
      }
    }, [disableInitialCall, autoRefresh, refreshInterval])

    return { call, ...state }
  }

  const useUpdateCall = <M extends keyof A>(
    args: ReActorUseUpdateArgs<A, M>
  ) => {
    return useReActorCall(args)
  }

  return {
    useQueryCall,
    useUpdateCall,
    useMethodField,
    useMethodFields,
  }
}
