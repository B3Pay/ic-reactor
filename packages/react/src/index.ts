import type { AuthClientLoginOptions } from "@dfinity/auth-client"
import type { ActorSubclass, ReActorOptions } from "@ic-reactor/store"
import { createReActorStore } from "@ic-reactor/store"
import { useEffect, useState } from "react"
import { useStore } from "zustand"
import { getCallHooks } from "./hooks"

export * from "./context"

export const createReActor = <A extends ActorSubclass<any>>(
  options: ReActorOptions
) => {
  const isLocal =
    typeof process !== "undefined" &&
    (process.env.NODE_ENV === "development" ||
      process.env.DFX_NETWORK === "local")

  const { callMethod, initialize, authenticate, authStore, actorStore } =
    createReActorStore<A>({
      isLocal,
      ...options,
    })

  const useActorStore = () => {
    const actorState = useStore(actorStore, (state) => state)

    return { ...actorState, initialize }
  }

  const useAuthStore = () => {
    const authState = useStore(authStore, (state) => state)
    return authState
  }

  const useAuthClient = () => {
    const [loginLoading, setLoginLoading] = useState(false)
    const [loginError, setLoginError] = useState<Error | unknown | null>(null)
    const { authClient, authenticated, authenticating, identity } =
      useAuthStore()

    const login = async (options?: AuthClientLoginOptions) => {
      setLoginLoading(true)
      setLoginError(null)

      try {
        await authClient?.login({
          ...options,
          onSuccess: async () => {
            setLoginLoading(false)
            await authenticate()
            options?.onSuccess?.()
          },
          onError: (e) => {
            setLoginError(e)
            setLoginLoading(false)
            options?.onError?.(e)
          },
        })
      } catch (e) {
        setLoginLoading(false)
        setLoginError(e)
      }
    }

    const logout = async (options?: { returnTo?: string }) => {
      await authClient?.logout(options)
      await authenticate()
    }

    useEffect(() => {
      if (!authClient && !authenticating) {
        authenticate()
      }
    }, [authClient, authenticating])

    return {
      authClient,
      authenticated,
      authenticating,
      identity,
      login,
      logout,
      authenticate,
      loginLoading,
      loginError,
    }
  }

  const { useQueryCall, useMethodField, useMethodFields, useUpdateCall } =
    getCallHooks<A>(callMethod, actorStore)

  return {
    useMethodFields,
    useMethodField,
    useActorStore,
    useAuthStore,
    useQueryCall,
    useUpdateCall,
    useAuthClient,
  }
}
