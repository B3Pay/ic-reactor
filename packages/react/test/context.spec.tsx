import React from "react"
import renderer, { act } from "react-test-renderer"
import { backend } from "./candid"
import { AgentProvider, createActorContext } from "../src"

const { ActorProvider, useQueryCall } = createActorContext<typeof backend>({
  canisterId: "xeka7-ryaaa-aaaal-qb57a-cai",
})

describe("createReactor", () => {
  it("should query", async () => {
    const TestComponent = ({}) => {
      const {
        call,
        data: version,
        loading,
      } = useQueryCall({ functionName: "version", refetchOnMount: false })

      return (
        <div>
          <button onClick={call}>Get Version</button>
          <span>
            {loading
              ? "Loading..."
              : version
              ? version.toString()
              : "Ready To call"}
          </span>
        </div>
      )
    }

    let screen = renderer.create(
      <AgentProvider>
        <ActorProvider>
          <TestComponent />
        </ActorProvider>
      </AgentProvider>
    )

    expect(screen.toJSON()).toMatchSnapshot()

    await act(async () => {
      await new Promise((r) => setTimeout(r, 3000))
    })

    expect(screen.toJSON()).toMatchSnapshot()

    const versionStatus = () => screen.root.findAllByType("span")[0]

    const versionButton = () => screen.root.findAllByType("button")[0]

    expect(versionStatus().props.children).toEqual("Ready To call")

    await act(() => versionButton().props.onClick())

    expect(versionStatus().props.children).toEqual("0.2.0")

    expect(screen.toJSON()).toMatchSnapshot()
  })
})
