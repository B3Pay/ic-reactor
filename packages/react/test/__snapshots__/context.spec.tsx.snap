// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createReactor should query 1`] = `
<div>
  Authenticating...
</div>
`;

exports[`createReactor should query 2`] = `
<div>
  <button
    onClick={[Function]}
  >
    Get Version
  </button>
  <span>
    Ready To call
  </span>
</div>
`;

exports[`createReactor should query 3`] = `
<div>
  <button
    onClick={[Function]}
  >
    Get Version
  </button>
  <span>
    0.2.0
  </span>
</div>
`;
